### ISSUE TITLE
The playlist can't show the Chinese filename correctlly! # Bug

### ISSUE THREAD

**Issue by huyuansong** [iina/iina#4682]
**System and IINA version:**

- macOS  :14.1 (23B74) 
- IINA  : 1.3.3 Build 138 

**Expected behavior:**
The file name in the finder

**Actual behavior:**
A wrong file name . It shows the description message of a file, maybe. Such as " record by EV 3.8.4 software" 

<details>
<summary><b>Crash report:</b></summary>

```
```
</details>

<br>

<details>
<summary><b>mpv log:</b></summary>

```
```
</details>

<br>

**Steps to reproduce:**

- [ ] MPV does not have this problem.

**How often does this happen?**

**Comment by svobs** [comment]
Are you playing an audio file? Looks like audio metadata.

Check this in Settings > General:
<img width="634" alt="image" src="https://github.com/iina/iina/assets/2213815/46924702-085e-4110-8779-c9529caf456d">

**Comment by huyuansong** [comment]
No, I am playing a video file. The player can show correctly when the file is not played. But It shows wrong file name when the file has been played.
Here is the picture.

￼

**Comment by huyuansong** [comment]
No, I am playing a video file. The player can show correctly when the file is not played. But It shows the wrong file name when the file has been played.
Here is the picture.
<img width="962" alt="QQ20231109-135415@2x-1" src="https://github.com/iina/iina/assets/19368111/3c623c09-bc90-475c-9661-01b9d162826f">

**Comment by svobs** [comment]
Interesting. I was able to produce strange behavior, but mine looks different.

<img width="424" alt="SCR-20231108-tyfa" src="https://github.com/iina/iina/assets/2213815/4a5703b4-5abd-487f-9460-b8ec47d80686">

I opened the first file in my folder, and then uses Next Track to move through all the tracks in my playlist. The title of the 2nd one changed when I moved past it. But none of the others did. Something is wrong there.

In your picture, it looks like the first 2 files have the same name now. Does it do that for every file? Also, is there anywhere else you see "record by EV 3.8.4 software" (in Chinese)? The folder name maybe? It must be getting that from somewhere...

IINA has complicated logic for changing the names in the playlist when it thinks it is part of a group. And there is a special logic for Chinese numbers. I don't completely understand it yet. But looks like there is a bug in there.

**Comment by huyuansong** [comment]
Some files can be shown correctly while others can not! All of the files in the playlist are the same thing. Those files are recorded by a screen recorder software as EV 3.8.4 (in Chinese). I guess that it is the one item of the metadata of those video files, maybe. I am sure that it does not come from the folder name in my case.

**Comment by svobs** [comment]
Found it. You are right. If the `title` metadata exists, it uses that, and if not, uses the filename.

The reason it doesn't use `title` right away is because it's asking the mpv library for that info, and mpv is initially returning null. It asks again each time you navigate through the playlist. Not exactly sure why that is. It's possible that mpv doesn't even try to load a file's metadata until it is opened for playing.

I think the obvious solution is to change the code not to use `title`. In my experience it's almost never used, and even when it is, it's usually junk like an ad for the encoding software.

**Comment by huyuansong** [comment]
OK, Fine! Thank you for checking it.

**Comment by low-batt** [comment]
@huyuansong Thank you for reporting this issue. I'm relatively new to IINA and was unaware of this playlist behavior. An additional thank you for closing this issue. It is helpful when users update their issues. However in this case I am going to reopen this issue and label it as a feature request.

The issue has to do with how IINA makes use of these two  [mpv](https://github.com/mpv-player/mpv) properties:
- [playlist/N/filename](https://mpv.io/manual/stable/#command-interface-playlist/n/filename):
	> Filename of the Nth entry.
- [playlist/N/title](https://mpv.io/manual/stable/#command-interface-playlist/n/title):
	> **playlist/N/title**
	> Name of the Nth entry. Available if the playlist file contains such fields and mpv's parser supports it for the given playlist format, or if the playlist entry has been opened before and a media-title other then then filename has been acquired.

Seems the manual has a typo that needs to be corrected.

The reference to [media-title](https://mpv.io/manual/stable/#command-interface-media-title) refers to this `mpv` property:
> **media-title**
>  If the currently played file has a title tag, use that.
>
>  Otherwise, return the filename property.

Notice the title can come from two places. The playlist itself can specify the title. The following is an example of an [Extended M3U](https://en.wikipedia.org/wiki/M3U#Extended_M3U) playlist:
```text
#EXTM3U
#EXTINF:154,The World in HDR in 4K
example-1.mp4
#EXTINF:634,Big Buck Bunny
example-2.mp4
#EXTINF:0,Big Buck Bunny Streaming
https://www.youtube.com/watch?v=aqz-KE-bpKQ
```
[example.m3u8.txt](https://github.com/iina/iina/files/13326278/example.copy.m3u8.txt)

When I open this playlist in IINA the playlist panel displays the titles contained in this playlist file.

This explains why IINA shows the title instead of the filename in the playlist panel. Embedded titles are a playlist feature.

Of course if the playlist does not provide a title then `mpv` falls back to a title tag in the media (if the media provides one), which `mpv` does not have access to until the media is opened.

What is missing is a way to view the filenames/URLs in the playlist if titles are available. This issue shows that unfortunately some software abuses the media title tag. That adds to the need for a way to switch the display from titles to filenames.

I'm not a UI designer. Hopefully @svobs or @uiryuu have some ideas on how to provide the ability to customize what is displayed in the playlist.

**Comment by svobs** [comment]
Unfortunate that this is turning into such a complex problem. I don't use playlist files or create playlists from network files, so those weren't cases I considered...

My thought: it's not good UX for items in the playlist should not change their names when opened. Their names should all be set immediately when the playlist is loaded, or shortly afterwards.

So there seems to be a desire by some to always use the title field, and others do not want to use the title field. As you noted, the obvious thing would be to add a user pref for that.

Looks like there is a bigger problem when using the title tags, because mpv only provides these when the media is opened. 

There is already a mechanism for loading artist & title asynchronously after the playlist item has been first displayed, then updating the item's name after they are retrieved. A similar strategy might be good for loading title meta, since it will involve opening each file to inspect it. But while that should at least be possible for local files, but I don't know if it is feasible for network files.

**Comment by low-batt** [comment]
For what to display for the window's title a preference works, but for the playlist you might want to normally see the titles but quickly switch to a view of the filenames. So I'm thinking a preference would control the default, but some way of quickly toggling the view is needed. What I was unsure of was where we'd put a button on the playlist panel, what that button would look it, how would this be added to the menu system, that kind of UI design.

And yes, a further enhancement possibly could include a background mechanism that populates the title, if the playlist did not supply it. Pretty sure we'd only want to do that for local files. Not a good idea to irritate streaming sites with lots of requests. That can cause the site to throttle you or even block your IP.

**Comment by svobs** [comment]
> Not a good idea to irritate streaming sites with lots of requests. That can cause the site to throttle you or even block your IP.

I took a look at how some other players handle this. For one thing, it seems that it is common for music players to update a streaming item's title once it starts playing, although I still don't like that. Fortunately I don't see them doing that for local files.

I did a quick search but had trouble finding other video players which support playlists. VLC does, and it seems to employ its own logic which changes the displayed name after a brief delay, suggesting that it is opening each for whatever reason. For the items in my folder, it looks like it decided to use the parent folder's name and the episode number from each filename, ignoring both the title field in the meta and the episode title in the filename. I don't think I'd try to emulate that.

> And yes, a further enhancement possibly could include a background mechanism that populates the title, if the playlist did not supply it. Pretty sure we'd only want to do that for local files.

From what I can gather, it looks like mpv doesn't really distinguish between a title which originated from a playlist file vs the meta from a media file. Which is unfortunate, because I'd otherwise suggest that we honor only the former.

> but for the playlist you might want to normally see the titles but quickly switch to a view of the filenames. So I'm thinking a preference would control the default, but some way of quickly toggling the view is needed. What I was unsure of was where we'd put a button on the playlist panel, what that button would look it, how would this be added to the menu system, that kind of UI design.

Yes, such a button wouldn't really fit with the existing buttons on the bottom, and might not be used often enough to justify putting it there...

The playlist context menu already provides 3 playlist-spanning operations at the bottom. How about adding something like a submenu with a choice of one checked item:
```
Display > 
   Filenames only
 ✓ Titles (if available)
```

**Comment by svobs** [comment]
> The playlist context menu already provides 3 playlist-spanning operations at the bottom. How about adding something like a submenu with a choice of one checked item:

Another thought, since I see that the context menu items are all represented at the bottom of the playlist. Could use an "eye" icon for the item, and clicking it would show a popup with the two display modes. Seems that the `+` button already displays a popup.

**Comment by jiangyixiang** [comment]
Same problem, finally decided to use v1.3.0, which displays the title correctly
